base/src/cache.rs

```diff
@@ -257,37 +313,549 @@ impl<T: Backend> CacheUpdate<T> for ChannelDelete {
+            futures.push(cache.members.upsert(member_entity));
+
+            let user_entity = UserEntity::from(member.user.clone());
+            futures.push(cache.users.upsert(user_entity));
+        }
+
+        for presence in self.presences.values() {
+            let entity = PresenceEntity::from(presence.clone());
+            futures.push(cache.presences.upsert(entity));
+        }
+
+        for role in self.roles.values() {
+            let entity = RoleEntity::from((role.clone(), self.id));
+            futures.push(cache.roles.upsert(entity));
+        }
+
+        for voice_state in self.voice_states.values() {
+            let entity = VoiceStateEntity::from((voice_state.clone(), self.id));
+            futures.push(cache.voice_states.upsert(entity));
+        }
+
+        let entity = GuildEntity::from(self.0.clone());
+        futures.push(cache.guilds.upsert(entity));
+
+        futures.try_collect().boxed()
+    }
+}
+
+impl<T: Backend + Send + Sync> CacheUpdate<T> for GuildDelete {
+    fn process<'a>(
+        &'a self,
+        cache: &'a Cache<T>,
+    ) -> Pin<Box<dyn Future<Output = Result<(), T::Error>> + Send + 'a>> {
+        if self.unavailable {
+            return cache
+                .guilds
+                .get(self.id)
+                .and_then(move |guild| {
+                    guild.map_or_else(
+                        || future::ok(()).boxed(),
+                        |guild| {
+                            let entity = GuildEntity {
+                                unavailable: self.unavailable,
+                                ..guild
+                            };
+
+                            cache.guilds.upsert(entity)
+                        },
+                    )
+                })
+                .boxed();
+        }
+
+        Box::pin(async move {
+            let futures = FuturesUnordered::new();
+
+            let mut channels = cache.guilds.channels(self.id).await?;
+            while let Some(Ok(c)) = channels.next().await {
+                match c {
+                    GuildChannelEntity::Category(c) => {
+                        futures.push(cache.category_channels.remove(c.id));
+                    }
+                    GuildChannelEntity::Text(c) => futures.push(cache.text_channels.remove(c.id)),
+                    GuildChannelEntity::Voice(c) => futures.push(cache.voice_channels.remove(c.id)),
+                }
+            }
+
+            let mut emojis = cache.guilds.emoji_ids(self.id).await?;
+            while let Some(Ok(id)) = emojis.next().await {
+                futures.push(cache.emojis.remove(id));
+            }
+
+            let mut members = cache.guilds.member_ids(self.id).await?;
+            while let Some(Ok(id)) = members.next().await {
+                futures.push(cache.members.remove((self.id, id)));
+            }
+
+            let mut presences = cache.guilds.presence_ids(self.id).await?;
```

Comment:
this can probably be some sort of try map or filter map to make this one statement instead of two and a loop

Reply:
I can't figure this one out; `TryStreamExt` doesn't actually have a `try_map` method, which is weird.